compiler/rustc_privacy/src/lib.rs

```diff
@@ -626,14 +643,30 @@ impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
     }
 
     fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
-        let item_level = match item.kind {
+        let item_ev = match item.kind {
             hir::ItemKind::Impl { .. } => {
-                let impl_level = Option::<Level>::of_impl(
+                let impl_ev = Option::<EffectiveVisibility>::of_impl(
                     item.owner_id.def_id,
                     self.tcx,
                     &self.effective_visibilities,
                 );
-                self.update(item.owner_id.def_id, impl_level)
+
+                if let Some(impl_ev) = impl_ev {
+                    self.changed |= self.effective_visibilities.update(
+                        item.owner_id.def_id,
+                        None,
+                        || {
+                            ty::Visibility::Restricted(
+                                self.tcx.parent_module_from_def_id(item.owner_id.def_id),
+                            )
+                        },
+                        impl_ev,
+                        Level::Direct,
+                        self.tcx,
+                    );
+                }
```

Comment:
Isn't this exactly the same logic as `fn update_eff_vis` is using?
Could it be reused here?